src/IECoreDelight/ShaderNetworkAlgo.cpp

```diff
@@ -0,0 +1,287 @@
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "Gaffer/Private/IECorePreview/LRUCache.h"
+
+#include "IECoreDelight/ShaderNetworkAlgo.h"
+
+#include "IECoreScene/ShaderNetworkAlgo.h"
+
+#include "IECore/SearchPath.h"
+#include "IECore/SimpleTypedData.h"
+#include "IECore/SplineData.h"
+#include "IECore/VectorTypedData.h"
+
+#include "OSL/oslquery.h"
+// Don't let windows.h smash over IECore::SearchPath via a macro
+#ifdef SearchPath
+#undef SearchPath
+#endif
+
+#include "boost/algorithm/string/predicate.hpp"
+
+using namespace Imath;
+using namespace IECore;
+using namespace IECoreScene;
+
+namespace
+{
+
+/////////////////////////////////////////////////////////////////////////
+// LRUCache of OSLQueries
+//////////////////////////////////////////////////////////////////////////
+
+using OSLQueryPtr = std::shared_ptr<OSL::OSLQuery>;
+using QueryCache = IECorePreview::LRUCache<std::string, OSLQueryPtr, IECorePreview::LRUCachePolicy::Parallel>;
+
+QueryCache &queryCache()
+{
+	static QueryCache g_cache(
+		[] ( const std::string &shaderName, size_t &cost, const IECore::Canceller *canceller ) -> OSLQueryPtr {
+			const char *searchPath = getenv( "OSL_SHADER_PATHS" );
+			OSLQueryPtr query = std::make_shared<OSL::OSLQuery>();
+			if( !query->open( shaderName, searchPath ? searchPath : "" ) )
+			{
+				return nullptr;
+			}
+			cost = 1;
```

Comment:
We need to assign to `cost` before any return statements, even the `return nullptr` (LRUCache doesn't have any special handling for `nullptr`, in fact it doesn't even require the value to be a pointer).

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/5432/commits/29ac7ebc75a1c26cb5955e501aa97b375a6f599c